pkg/controller/perconaservermongodb/finalizers.go

```diff
@@ -2,32 +2,77 @@ package perconaservermongodb
 
 import (
 	"context"
-
 	api "github.com/percona/percona-server-mongodb-operator/pkg/apis/psmdb/v1"
 	"github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
-func (r *ReconcilePerconaServerMongoDB) checkFinalizers(cr *api.PerconaServerMongoDB) error {
+func (r *ReconcilePerconaServerMongoDB) checkFinalizers(cr *api.PerconaServerMongoDB) (shouldReconcile bool, err error) {
+	shouldReconcile = true
 	finalizers := []string{}
 
 	for _, f := range cr.GetFinalizers() {
 		switch f {
 		case "delete-psmdb-pvc":
-			err := r.deletePvcFinalizer(cr)
+			err = r.deletePvcFinalizer(cr)
+		case "delete-psmdb-pods-in-order":
+			err = r.deletePSMDBPods(cr)
 			if err != nil {
-				log.Error(err, "failed to run finalizer", "finalizer", f)
-				finalizers = append(finalizers, f)
+				shouldReconcile = false
 			}
 		}
+		if err != nil {
+			log.Error(err, "failed to run finalizer", "finalizer", f)
+			finalizers = append(finalizers, f)
+		}
 	}
-
 	cr.SetFinalizers(finalizers)
-	err := r.client.Update(context.TODO(), cr)
+	err = r.client.Update(context.TODO(), cr)
+
+	return shouldReconcile, err
+}
+
+func (r *ReconcilePerconaServerMongoDB) deletePSMDBPods(cr *api.PerconaServerMongoDB) (err error) {
+	done := true
+	for _, rs := range cr.Spec.Replsets {
+		sts, err := r.getRsStatefulset(cr, rs.Name)
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return errors.Wrap(err, "get rs statefulset")
+		}
+		if k8serrors.IsNotFound(err) {
+			continue
+		}
 
-	return err
+		pods := &corev1.PodList{}
+		err = r.client.List(context.TODO(),
+			pods,
+			&client.ListOptions{
+				Namespace:     cr.Namespace,
+				LabelSelector: labels.SelectorFromSet(sts.Labels),
+			},
+		)
+		if err != nil && !k8serrors.IsNotFound(err) {
+			return errors.Wrap(err, "get rs statefulset")
+		}
+		if k8serrors.IsNotFound(err) {
+			continue
+		}
+		if len(pods.Items) > int(*sts.Spec.Replicas) {
+			return errors.New("waiting pods to be deleted")
+		}
+		if *sts.Spec.Replicas != 1 {
+			rs.Size = 1
+			done = false
```

Comment:
If there are multiple replicasets, we may overwrite value of `done`

Reply:
That's correct. We cycle across every replicaset, and if one of them has size > 1, then we overwrite done to call `deletePSMDBPods` on the next reconcile loop